### PR TITLE
Equiv

### PR DESCRIPTION
**Pull request by zhanglw0521**

### CODE REVIEW

**Comment by cortner** [comment]
Did you make the PR on purpose or by accident? It is definitely not ready to be merged. Not a problem really, we can continue to push to your branch until it is ready for merging.

**Comment by cortner** [comment]
Step one might be to merge the latest `ACEsuit:equiv` into  your new branch. I've made quite a few changes in the interim.

**Comment by cortner** [comment]
I'll quickly grab some lunch then we could continue on this if you are free.

**Comment by cortner** [comment]
tests are now running, but don't pass yet.

**Comment by cortner** [comment]
@zhanglw0521 I've just merged Matthias' PR for euclidean vectors. This will mean extra work for us to make his changes compatible with yours. I'll see if I can do it at my end, but we might need a meeting amongst the three of us. cf. ZuLIP.

**Comment by cortner** [comment]
@zhanglw0521 I've changed the base branch to `main` since this is where we are now working.

**Comment by cortner** [comment]
looks like there are still conflict we will need to fix before we can merge. strange

**Comment by zhanglw0521** [comment]
> looks like there are still conflict we will need to fix before we can merge. strange

I think the conflict in ACE.jl is caused by the rearrangement of my code - I've moved function `Rotation_D_matrix` and `Rotation_D_matrix_ast` to module `Wigner` so that we may need `Wigner` before using `Rotations3D`.

Re conflict in properties.jl/rotations3d-equiv.jl, it seems that the latest one (which is now in `main` branch) is the correct one?

File: src/properties.jl [hunk 1]
```diff
@@ -8,6 +8,8 @@ abstract type AbstractProperty end
 @inline -(φ1::T, φ2::T) where {T <: AbstractProperty} = T( φ1.val - φ2.val )
 @inline -(φ::T) where {T <: AbstractProperty} = T( -φ.val)
 @inline *(φ::T, λ::Number) where {T <: AbstractProperty} = T(φ.val * λ)
+@inline *(a::Union{Number, AbstractMatrix}, φ::T) where {T <: AbstractProperty} =
```
Comment (cortner) [inline src/properties.jl:11]: lines 19 and 20 below seem to be no longet necessary - please delete

Comment (cortner) [reply src/properties.jl:11]: i.e. delete these two:
```
@inline *(A::AbstractMatrix, φ::T) where {T <: AbstractProperty} = T(A * φ.val)
# @inline *(A::StaticArrays.SArray{Tuple{3,3}, T,2,9}, φ::EuclideanVector{T}) where {T <: Number} = EuclideanVector{T}(A * φ.val)
```

File: src/properties.jl [hunk 2]
```diff
@@ -89,22 +91,56 @@ end
 
 struct SphericalVector{L, LEN, T} <: AbstractProperty
    val::SVector{LEN, T}
-   _valL::Val{L}
+   _valL::Val{L} ## Why do we need Val?
```
Comment (cortner) [inline src/properties.jl:94]: probably we don't. I'm happy to explore this later. Can we keep it for now please?

File: src/properties.jl [hunk 3]
```diff
@@ -89,22 +91,56 @@ end
...
 end
 
+getL(φ::SphericalVector) = typeof(φ).parameters[1];
```
Comment (cortner) [inline src/properties.jl:97]: replace with 
```julia
getL(φ::SphericalVector{L}) where {L} = L 
```

File: src/properties.jl [hunk 4]
```diff
@@ -89,22 +91,56 @@ end
...
+SphericalVector{L, LEN, T}()  where {L, LEN, T} =
       SphericalVector( zero(SVector{LEN, T}), Val{L}() )
 
+filter(φ::SphericalVector, b::Array) = ( length(b) <= 1 ? true :
+        ( ( iseven(sum(bi.l for bi in b)) == iseven(getL(φ)) ) &&
+         ( abs(sum(bi.m for bi in b)) <= getL(φ) )  ) )
+
+rot3Dcoeffs(::SphericalVector, T::DataType=Float64) = Rot3DCoeffs(T)
+
+struct Sphericalvector{LEN, T} <: AbstractProperty
```
Comment (cortner) [inline src/properties.jl:123]: Why are there still `SphericalVector` and `Sphericalvector`?

File: src/rotations3d-spher-vec.jl [hunk 5]
```diff
@@ -0,0 +1,145 @@
+
+# Index of entries in D matrix (sign included)
+struct D_Index
+	sign::Int64
+	μ::Int64
+	m::Int64
+end
+
+# Equation (1.1) - forms the covariant matrix D(Q)(indices only)
+function Rotation_D_matrix(L::Integer)
```
Comment (cortner) [inline src/rotations3d-spher-vec.jl:10]: functions should start with lower-case, please rename 
```
rotation_D_matrix
```

File: src/rotations3d-spher-vec.jl [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    for i = 1 : 2 * L + 1
+        for j = 1 : 2 * L + 1
+            D[j,i] = D_Index(1, i - 1 - L, j - 1 - L);
+        end
+    end
+	return D
+end
+
+# Equation (1.1) - forms the covariant matrix D(Q)(indices only)
+function Rotation_D_matrix_ast(L::Integer)
```
Comment (cortner) [inline src/rotations3d-spher-vec.jl:24]: same as above, rename
```
rotation_D_matrix_ast
```

File: src/rotations3d-spher-vec.jl [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+	rk = rank(G; rtol =  1e-8)
+	Urcpi = fill(zeros(2L + 1), (rk, length(μ_list)))
+	U = S.U[:, 1:rk]
+	Sigma = S.S[1:rk]
+	Urcpi = C * U * Diagonal(sqrt.(Sigma))^(-1)
+	return Urcpi', μ_list
+end
+
+# Equation (1.12) - Gramian over nn
+function Gramian(A::Rot3DCoeffs,
```
Comment (cortner) [inline src/rotations3d-spher-vec.jl:111]: lower case please

File: test/testing/test_wigner.jl [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+using ACE: evaluate
+using Random, Printf, Test, ACE.Testing
+using ACE, Test, Printf, LinearAlgebra, BenchmarkTools
+
+##
+
+
+@info("Check correctness of Wigner-D matrices")
+
+L = 1
```
Comment (cortner) [inline test/testing/test_wigner.jl:17]: Can you please generalise this test to allow `for L = 0:3` as in the other test?

**Review by cortner — CHANGES_REQUESTED** [review]
Hi Liwei - thanks for your work on this so far. Can you fix a few more things as discused in this review please? Any questions you can just respond here in the review. Longer discussion on ZuLIP I think.

File: src/properties.jl (hunk 2 above)
Comment (zhanglw0521) [reply src/properties.jl:94]: We do need this as discussed. I just forgot to remove the comment, sorry..

**Pull request merged**